rust-runtime/aws-smithy-async/src/test_util/tick_advance_sleep.rs

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Test time/sleep implementations that work by manually advancing time with a `tick()`
+//!
+//! # Examples
+//!
+//! Spawning a task that creates new sleep tasks and waits for them sequentially,
+//! and advancing passed all of them with a single call to `tick()`.
+//!
+//! ```rust,no_run
+//! use std::time::{Duration, SystemTime};
+//! use aws_smithy_async::test_util::tick_advance_sleep::tick_advance_time_and_sleep;
+//! use aws_smithy_async::time::TimeSource;
+//! use aws_smithy_async::rt::sleep::AsyncSleep;
+//!
+//! # async fn example() {
+//! // Create the test time/sleep implementations.
+//! // They will start at SystemTime::UNIX_EPOCH.
+//! let (time, sleep) = tick_advance_time_and_sleep();
+//!
+//! // Spawn the task that sequentially sleeps
+//! let task = tokio::spawn(async move {
+//!     sleep.sleep(Duration::from_secs(1)).await;
+//!     sleep.sleep(Duration::from_secs(2)).await;
+//!     sleep.sleep(Duration::from_secs(3)).await;
+//! });
+//! // Verify that task hasn't done anything yet since we haven't called `tick`
+//! tokio::task::yield_now().await;
+//! assert!(!task.is_finished());
+//! assert_eq!(SystemTime::UNIX_EPOCH, time.now());
+//!
+//! // Tick 6 seconds, which is long enough to go passed all the sequential sleeps
+//! time.tick(Duration::from_secs(6)).await;
+//! assert_eq!(SystemTime::UNIX_EPOCH + Duration::from_secs(6), time.now());
+//!
+//! // Verify the task joins, indicating all the sleeps are done
+//! task.await.unwrap();
+//! # }
+//! ```
+
+use crate::{
+    rt::sleep::{AsyncSleep, Sleep},
+    time::TimeSource,
+};
+use std::{
+    future::IntoFuture,
+    ops::{Deref, DerefMut},
+    sync::{Arc, Mutex},
+    time::{Duration, SystemTime},
+};
+use tokio::sync::oneshot::Sender;
+
+#[derive(Debug)]
+struct QueuedSleep {
+    /// Duration since `UNIX_EPOCH` at which point the sleep is finished.
+    presents_at: Duration,
+    notify: Option<Sender<()>>,
+}
+
+#[derive(Default, Debug)]
+struct Inner {
+    sleeps: Vec<QueuedSleep>,
+    /// Duration since `UNIX_EPOCH` that represents "now".
+    now: Duration,
+}
+
+impl Inner {
+    fn take_presenting_before(&mut self, time: Duration) -> Vec<QueuedSleep> {
+        self.sleeps.sort_by_key(|s| s.presents_at);
```

Comment:
Does `self.sleeps` need be a `Vec` if we need to sort elements upon `take_presenting_before` being called? Can something like `BTreeMap` be an option?

Reply:
Added a comment about the collection choice in c492b4ee9ae5d0ea198dbe6979bc65fbbe920acd